File: compiler/src/codegen/mangle.rs
use crate::{codegen::ir, database::NodeRef};
use std::fmt::Write as _;

pub trait Mangle {
    fn mangle(&self) -> String;
}

impl Mangle for NodeRef {
    fn mangle(&self) -> String {
        format!("_{}", self.id())
    }
}

impl Mangle for ir::DefinitionKey {
    fn mangle(&self) -> String {
        let mut s = self.node.mangle();

        for (parameter, ty) in &self.substitutions {
            write!(s, "_{}_{}", parameter.id(), ty.mangle_nominal()).unwrap();
        }

        // Bounds do not need to be part of the mangled ID because instances are
        // globally unique

        s
    }
}

impl ir::Type {
    pub fn mangle_nominal(&self) -> String {
        self.mangle_inner(true).unwrap()
    }

    pub fn mangle_structural(&self) -> Option<String> {
        self.mangle_inner(false)
    }

    fn mangle_inner(&self, nominal: bool) -> Option<String> {
        match self {
            ir::Type::Named(node, parameters, flags) => {
                if !nominal && flags.intrinsic {
                    None
                } else {
                    let mut s = format!("type{}", node.mangle());

                    for parameter in parameters {
                        write!(s, "_{}", parameter.mangle_nominal()).unwrap();
                    }

                    Some(s)
                }
            }
            ir::Type::Tuple(elements) => {
                let mut s = String::from("tuple");

                for element in elements {
                    write!(s, "_{}", element.mangle_nominal()).unwrap();
                }

                Some(s)
            }
            ir::Type::Function(inputs, output) => {
                let mut s = String::from("function");

                for input in inputs {
                    write!(s, "_{}", input.mangle_nominal()).unwrap();
                }

                write!(s, "_{}", output.mangle_nominal()).unwrap();

                Some(s)
            }
            ir::Type::Parameter(node) => Some(format!("parameter{}", node.mangle())),
        }
    }
}

File: compiler/src/codegen/Mangle.swift
protocol Mangleable { var mangled: String { get } }

extension Node: Mangleable { var mangled: String { "_\(self.id)" } }

extension IR.DefinitionKey: Mangleable {
    var mangled: String {
        var mangled = self.node.mangled

        for (parameter, type) in self.substitutions {
            mangled += "\(parameter.mangled)_\(type.nominal)"
        }

        // Bounds do not need to be part of the mangled ID because instances are
        // globally unique

        return mangled
    }
}

extension IR.`Type` {
    var nominal: String { self.mangled(nominal: true)! }

    var structural: String? { self.mangled(nominal: false) }

    private func mangled(nominal: Bool) -> String? {
        switch self {
        case .named(let definition, let parameters, let intrinsic, let representation, _):
            guard nominal || (!intrinsic && representation == nil) else { return nil }

            var mangled = "type\(definition.mangled)"
            for parameter in parameters { mangled += "_\(parameter.nominal)" }
            return mangled
        case .tuple(let elements):
            var mangled = "tuple"
            for element in elements { mangled += "_\(element.nominal)" }
            return mangled
        case .function(let inputs, let output):
            var mangled = "function"
            for input in inputs { mangled += "_\(input.nominal)" }
            mangled += "_\(output.nominal)"
            return mangled
        case .parameter(let definition): return "parameter\(definition.mangled)"
        }
    }
}
